### ISSUE TITLE
[Epic] Rosbag2 playback time control and simulation time

### ISSUE THREAD

**Issue by emersonknapp** [ros2/rosbag2#696]
## Description

Tracking issue for the group of features (planned for Galactic) around control of the flow of time during rosbag2 playback. See "Completion Criteria" for full feature list

## Design

* [x]  Initial Design: https://github.com/ros2/rosbag2/pull/675
* [x] Extension with time conversion logic #700 
* [x] Extension with `play_next` https://github.com/ros2/rosbag2/pull/708

## Feature List / Completion Criteria

Design at https://github.com/ros2/rosbag2/blob/master/docs/design/rosbag2_playback_time.md is fully implemented including all of the following features (names tagged where items are known to be work-in-progress):

* [x] `PlayerClock` integration into `Player` (@emersonknapp)
    * [X] (bugfix/unblock) Fix incorrect filter test: #690
    * [X] Initial `PlayerClock` implementation: #689
    * [x] Thread safety annotation followup #710
* [x] pause/resume time during playback #55 (@emersonknapp)
    * [x] impl: #704 
    * [x] fix behavior https://github.com/ros2/rosbag2/pull/730
    * [x] add error checking https://github.com/ros2/rosbag2/pull/731
* [x] change rate during playback (@emersonknapp )
    * [x] impl: #727
* [x] jump/seek to arbitrary time during playback (@MichaelOrlov) #821
    * [x] naive clock jump implementation for simpler setup https://github.com/ros2/rosbag2/pull/754 (@emersonknapp)
    * [x] jump callbacks #775 (@MichaelOrlov)
    * [x] https://github.com/ros2/rosbag2/issues/821
    * [x] storage seek API #824 (@lihui815)
        * [x] #836 
        * [x] https://github.com/ros2/rosbag2/pull/873
* [x] "Play next message" while paused (@MichaelOrlov) https://github.com/ros2/rosbag2/pull/762
* [x] pause/rate/play-next controllable by keyboard handler 
    * [x] #55 (@MichaelOrlov)
    * [x] Design document for unified keyboard handler https://github.com/ros2/rosbag2/pull/768
    * [x] Unified keyboard handler https://github.com/ros2/rosbag2/pull/735 (@MichaelOrlov)
    * [x] expose Rate to keyboard
* [x] pause/rate/play-next/jump controllable by external caller via Service (@emersonknapp )
    * [x] Add service definitions https://github.com/ros2/rosbag2/pull/728 
    * [x] Expose pause/resume services on the Player https://github.com/ros2/rosbag2/pull/729
    * [x] Expose playback rate services on the Player https://github.com/ros2/rosbag2/pull/753
    * [x] Expose play_next() API via service https://github.com/ros2/rosbag2/pull/767 (@MichaelOrlov)
    * [x] Expose jump service on the Player https://github.com/ros2/rosbag2/pull/874
* [x] can publish to `/clock` during playback #99 (@emersonknapp)
    * [x] Migrate `play` to `rosbag2_py` to make it easier to add new playback options: #693
    * [x] Clock Publisher: #695
* [ ] can subscribe to `/clock` during playback (`use-sim-time`) #694 (@emersonknapp )
    * [x] https://github.com/ros2/rosbag2/pull/752 
    * [ ] https://github.com/ros2/rclcpp/issues/1730
    * [ ] https://github.com/ros2/rclcpp/pull/1748

### Code cleanup / tech debt / rearchitecture

* [x] Migrate `record` to `rosbag2_py` #702
* [x] Test flakiness fix #705
* [x] Fix rosbag2_py after rclpy moved rmw_qos_profile_t to be a py::class #707
* [x] use `rclcpp`'s GenericPub/Sub, remove from rosbag2 (followup to https://github.com/ros2/rclcpp/pull/1452) (@Karsten1987) #714 
* [x] Split `Rosbag2Transport` into separate `Player` and `Recorder` classes #721
* [x] `Player` and `Recorder` as Nodes 
    * [x] #716
    * [x] https://github.com/ros2/rosbag2/pull/723
    * [x] https://github.com/ros2/rosbag2/pull/724
    * [x] https://github.com/ros2/rosbag2/pull/725
    * [x] https://github.com/ros2/rosbag2/pull/726

**Comment by ros-discourse** [comment]
This issue has been mentioned on **ROS Discourse**. There might be relevant details there:

https://discourse.ros.org/t/tooling-wg-breakout-session-rosbag2-playback-time-control/19711/1

**Comment by Karsten1987** [comment]
Once https://github.com/ros2/rclcpp/pull/1452 is in, I'd take the refactoring of removing the `rosbag2_node` and replace it with the new rclcpp one. Does this sound ok or do I disturb other people's workflow with this? 
I feel that the task list above might lead to quite a few conflicts as they massively overlap.

**Comment by emersonknapp** [comment]
I think that would be fine - I don't see it interfering very much with the existing line items. Most of the changes now should be very focused, and not be near the direct usage of `GenericPublisher`

**Comment by Karsten1987** [comment]
Ok, I was just afraid that I would overlap with @MichaelOrlov's last task: "Split Rosbag2Transport into separate Player and Recorder classes"

**Comment by emersonknapp** [comment]
Yes, that one could overlap, it would probably be good to stage those serially.

**Comment by emersonknapp** [comment]
But - maybe it won't be too bad! Perhaps you can stage it as "replace rosbag2::generic with rclcpp::generic" - which definitely will have little to no conflict - then do "remove the rosbag2node" as a second stage

**Comment by MichaelOrlov** [comment]
@Karsten1987 Yeah I think it would overlap. \
Please go ahead first. I can try to tackle another tasks for a while. Such as `jump`, `play_next` and keyboard handling.

**Comment by Kaju-Bubanja** [comment]
I don't see this awesome feature in the list of new features here: https://docs.ros.org/en/foxy/Releases/Release-Galactic-Geochelone.html#timeline-before-the-release Is it still planned to be released for galactic?

**Comment by emersonknapp** [comment]
Yes - everything that is currently checked off will be in Galactic (and maybe a couple more small things). None of the rosbag2 changes are listed on that page.

**Comment by clalancette** [comment]
> None of the rosbag2 changes are listed on that page.

I hope you mean "yet" :smile: .  It would be awesome to have a high-level overview of the (massive) changes to rosbag2 listed on that page.

**Comment by emersonknapp** [comment]
> I hope you mean "yet" smile . It would be awesome to have a high-level overview of the (massive) changes to rosbag2 listed on that page.

Yes, yet. Should we open a PR against the page source for it?

**Comment by clalancette** [comment]
> Yes, yet. Should we open a PR against the page source for it?

Yes, whenever you are ready, please feel free to open a PR and I'm happy to review it.

**Comment by emersonknapp** [comment]
Closing this issue, we don't need it anymore. It's a huge amount of work that we did! The only thing really remaining is use-sim-tim, which can be tracked via its own issue https://github.com/ros2/rosbag2/issues/694
